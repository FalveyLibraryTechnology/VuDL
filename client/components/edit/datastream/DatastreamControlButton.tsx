import React, { useState } from "react";
import PropTypes from "prop-types";
import LoadingButton from "@mui/lab/LoadingButton";
import Tooltip from "@mui/material/Tooltip";
import Download from "@mui/icons-material/Download";
import Delete from "@mui/icons-material/Delete";
import UploadFile from "@mui/icons-material/UploadFile";
import { useEditorContext } from "../../../context/EditorContext";
import useDatastreamOperation from "../../../hooks/useDatastreamOperation";

const Icons = {
    Upload: <UploadFile />,
    // VIEW: ,
    // METADATA: ,
    Download: <Download />,
    Delete: <Delete />,
};

const DatastreamControlButton = ({ modalState, datastream, disabled }) => {
    const [isLoading, setLoading] = useState(false);
    const {
        action: { toggleDatastreamModal, setActiveDatastream, setDatastreamModalState },
    } = useEditorContext();
    const { downloadDatastream } = useDatastreamOperation();
    const onClick = (modalState) => {
        if (modalState !== "Download") {
            return () => {
                setActiveDatastream(datastream);
                setDatastreamModalState(modalState);
                toggleDatastreamModal();
            };
        }
        return async () => {
            setLoading(true);
            await downloadDatastream(datastream);
            setLoading(false);
        };
    };

    return (
        <Tooltip title={modalState}>
            <LoadingButton
                className="datastreamControlButton"
                loading={isLoading}
                aria-label={modalState}
                disabled={modalState !== "Upload" && disabled}
                onClick={onClick(modalState)}
                size="small"
            >
                {Icons[modalState]}
            </LoadingButton>
        </Tooltip>
    );
};

DatastreamControlButton.propTypes = {
    modalState: PropTypes.string,
    disabled: PropTypes.bool,
    datastream: PropTypes.string,
};
export default DatastreamControlButton;
