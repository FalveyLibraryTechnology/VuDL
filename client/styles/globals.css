:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
