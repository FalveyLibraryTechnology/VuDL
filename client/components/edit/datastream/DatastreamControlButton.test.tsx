import React from "react";
import { describe, beforeEach, expect, it, jest } from "@jest/globals";
import { mount, shallow } from "enzyme";
import { act } from "react-dom/test-utils";
import toJson from "enzyme-to-json";
import DatastreamControlButton from "./DatastreamControlButton";

const mockUseEditorContext = jest.fn();
jest.mock("../../../context/EditorContext", () => ({
    useEditorContext: () => {
        return mockUseEditorContext();
    },
}));
const mockUseDatastreamOperation = jest.fn();
jest.mock("../../../hooks/useDatastreamOperation", () => () => mockUseDatastreamOperation());
describe("DatastreamControlButton", () => {
    let editorValues;
    let datastreamOperationValues;
    beforeEach(() => {
        editorValues = {
            action: {
                toggleDatastreamModal: jest.fn(),
                setActiveDatastream: jest.fn(),
                setDatastreamModalState: jest.fn(),
            },
        };
        datastreamOperationValues = {
            downloadDatastream: jest.fn(),
        };
        mockUseEditorContext.mockReturnValue(editorValues);
        mockUseDatastreamOperation.mockReturnValue(datastreamOperationValues);
    });

    it("renders", () => {
        const wrapper = shallow(<DatastreamControlButton modalState="Upload" datastream="THUMBNAIL" />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("downloads the datastream", async () => {
        const wrapper = mount(<DatastreamControlButton modalState="Download" datastream="THUMBNAIL" />);
        await act(async () => {
            wrapper.find("button.datastreamControlButton").simulate("click");
            wrapper.update();
        });
        expect(datastreamOperationValues.downloadDatastream).toHaveBeenCalledWith("THUMBNAIL");
    });

    it("activates the modal", async () => {
        const wrapper = mount(<DatastreamControlButton modalState="View" datastream="THUMBNAIL" />);
        await act(async () => {
            wrapper.find("button.datastreamControlButton").simulate("click");
            wrapper.update();
        });
        expect(editorValues.action.setActiveDatastream).toHaveBeenCalledWith("THUMBNAIL");
        expect(editorValues.action.setDatastreamModalState).toHaveBeenCalledWith("View");
        expect(editorValues.action.toggleDatastreamModal).toHaveBeenCalled();
    });
});
