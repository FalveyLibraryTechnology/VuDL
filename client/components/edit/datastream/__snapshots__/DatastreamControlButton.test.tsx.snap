// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatastreamControlButton renders 1`] = `
<ForwardRef(Tooltip)
  title="Upload"
>
  <ForwardRef(LoadingButton)
    aria-label="Upload"
    className="datastreamControlButton"
    disabled={false}
    loading={false}
    onClick={[Function]}
    size="small"
  >
    <Memo(ForwardRef(UploadFileIcon)) />
  </ForwardRef(LoadingButton)>
</ForwardRef(Tooltip)>
`;
